polkadot/xcm/xcm-builder/src/barriers.rs

```diff
@@ -200,18 +201,34 @@ impl<InnerBarrier: ShouldExecute, LocalUniversal: Get<InteriorLocation>, MaxPref
 							return Err(ProcessMessageError::Unsupported)
 						};
 					},
+					ExecuteWithOrigin { origin: Some(interior), xcm } => {
+						let Ok(_) = actual_origin.append_with(interior.clone()) else {
+							return Err(ProcessMessageError::Unsupported)
+						};
+						InnerBarrier::should_execute(
+							&actual_origin,
+							&mut xcm.inner_mut(),
+							max_weight,
+							properties,
+						)?;
+						already_checked_inner_barrier = true;
```

Comment:
It might be incomplete. This allows the use case of paying fees using `ExecuteWithOrigin` at the start. It doesn't allow recursion but it's debatable how useful that is.